File: record/android/src/main/java/com/llfbandit/record/methodcall/MethodCallHandlerImpl.java
package com.llfbandit.record.methodcall;

import androidx.annotation.NonNull;

import com.llfbandit.record.Utils;
import com.llfbandit.record.permission.PermissionManager;
import com.llfbandit.record.record.RecordConfig;
import com.llfbandit.record.stream.RecorderRecordStreamHandler;
import com.llfbandit.record.stream.RecorderStateStreamHandler;

import java.io.IOException;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class MethodCallHandlerImpl implements MethodCallHandler {
  private final PermissionManager permissionManager;
  private final RecorderWrapper recorderWrapper;

  public MethodCallHandlerImpl(
      @NonNull PermissionManager permissionManager,
      @NonNull RecorderStateStreamHandler recorderStateStreamHandler,
      @NonNull RecorderRecordStreamHandler recorderRecordStreamHandler) {

    this.permissionManager = permissionManager;

    recorderWrapper = new RecorderWrapper(recorderStateStreamHandler,recorderRecordStreamHandler);
  }

  public void close() {
    recorderWrapper.close();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method) {
      case "start":
        try {
          RecordConfig config = getRecordConfig(call, true);
          recorderWrapper.startRecordingToFile(config, result);
        } catch (IOException e) {
          result.error("record", "Cannot create recording configuration.", e.getMessage());
        }
        break;
      case "startStream":
        try {
          RecordConfig config = getRecordConfig(call, false);
          recorderWrapper.startRecordingToStream(config, result);
        } catch (IOException e) {
          result.error("record", "Cannot create recording configuration.", e.getMessage());
        }
        break;
      case "stop":
        recorderWrapper.stop( result);
        break;
      case "pause":
        recorderWrapper.pause( result);
        break;
      case "resume":
        recorderWrapper.resume( result);
        break;
      case "isPaused":
        recorderWrapper.isPaused( result);
        break;
      case "isRecording":
        recorderWrapper.isRecording( result);
        break;
      case "hasPermission":
        permissionManager.hasPermission(result::success);
        break;
      case "getAmplitude":
        recorderWrapper.getAmplitude( result);
        break;
      case "listInputDevices":
        result.success(null);
        break;
      case "dispose":
        close();
        result.success(null);
        break;
      case "isEncoderSupported":
        String codec = call.argument("encoder");
        boolean isSupported = AudioRecorder.isEncoderSupported(codec);
        isSupported |= MediaRecorder.isEncoderSupported(codec);
        result.success(isSupported);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  private RecordConfig getRecordConfig(@NonNull MethodCall call, boolean genPath) throws IOException {
    String path = null;

    if (genPath) {
      String encoder = Utils.firstNonNull(call.argument("encoder"), "aacLc");
      path = Utils.genTempFileName(encoder);
    }

    return new RecordConfig(
        path,
        Utils.firstNonNull(call.argument("encoder"), "aacLc"),
        Utils.firstNonNull(call.argument("bitRate"), 128000),
        Utils.firstNonNull(call.argument("samplingRate"), 44100),
        Utils.firstNonNull(call.argument("numChannels"), 2),
        call.argument("device"),
        Utils.firstNonNull(call.argument("noiseCancel"), false),
        Utils.firstNonNull(call.argument("autoGain"), false)
    );
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/MethodCallHandlerImpl.kt
package com.llfbandit.record

import android.content.Context
import com.llfbandit.record.permission.PermissionManager
import com.llfbandit.record.record.format.AudioFormats
import com.llfbandit.record.record.model.RecordConfig
import com.llfbandit.record.record.util.DeviceUtils
import com.llfbandit.record.record.util.MainThread
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class MethodCallHandlerImpl(
  private val permissionManager: PermissionManager,
  private val messenger: BinaryMessenger,
  private val appContext: Context,
) : MethodChannel.MethodCallHandler {
  private val recorders = HashMap<String, RecorderWrapper>()

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    val recorderId = call.argument<String>("recorderId")

    if (recorderId.isNullOrEmpty()) {
      result.error("record", "Call missing mandatory parameter recorderId.", null)
      return
    }

    // Stateless, or UI-bound (permission dialogs must run on the platform thread).
    when (call.method) {
      "hasPermission" -> { hasPermission(call, result); return }
      "listInputDevices" -> { result.success(DeviceUtils.listInputDevicesAsMap(appContext)); return }
      "isEncoderSupported" -> { isEncoderSupported(call, result); return }
    }

    if (call.method == "create") {
      createRecorder(recorderId, result)
      return
    }

    val recorder = recorders[recorderId]
    if (recorder == null) {
      result.error(
        "record",
        "Recorder has not yet been created or has already been disposed.", null
      )
      return
    }

    val mainResult = MainThreadResult(result)
    when (call.method) {
      "start" -> recorder.startRecordingToFile(RecordConfig.fromMap(call, appContext), mainResult)
      "startStream" -> recorder.startRecordingToStream(RecordConfig.fromMap(call, appContext), mainResult)
      "stop" -> recorder.stop(mainResult)
      "pause" -> recorder.pause(mainResult)
      "resume" -> recorder.resume(mainResult)
      "isPaused" -> recorder.isPaused(mainResult)
      "isRecording" -> recorder.isRecording(mainResult)
      "cancel" -> recorder.cancel(mainResult)
      "getAmplitude" -> recorder.getAmplitude(mainResult)
      "dispose" -> disposeRecorder(recorder, recorderId, mainResult)
      else -> result.notImplemented()
    }
  }

  fun dispose() {
    // Iterate a snapshot: disposeRecorder() mutates the backing map.
    for ((recorderId, recorder) in HashMap(recorders)) {
      disposeRecorder(recorder, recorderId, null)
    }
    recorders.clear()
  }

  // Marshals Result callbacks back to the platform thread Flutter requires.
  private class MainThreadResult(
    private val delegate: MethodChannel.Result
  ) : MethodChannel.Result {
    override fun success(result: Any?) { MainThread.post { delegate.success(result) } }
    override fun error(code: String, message: String?, details: Any?) {
      MainThread.post { delegate.error(code, message, details) }
    }
    override fun notImplemented() { MainThread.post { delegate.notImplemented() } }
  }

  private fun createRecorder(recorderId: String, result: MethodChannel.Result) {
    if (recorders.containsKey(recorderId)) {
      result.success(null)
      return
    }

    try {
      recorders[recorderId] = RecorderWrapper(appContext, recorderId, messenger)
      result.success(null)
    } catch (e: Exception) {
      result.error("record", "Cannot create recorder.", e.message)
    }
  }

  private fun disposeRecorder(recorder: RecorderWrapper, recorderId: String, result: MethodChannel.Result?) {
    recorder.dispose(result)
    recorders.remove(recorderId)
  }

  private fun hasPermission(call: MethodCall, result: MethodChannel.Result) {
    val request = call.argument<Boolean>("request") ?: true
    permissionManager.hasPermission(request, result::success)
  }

  private fun isEncoderSupported(call: MethodCall, result: MethodChannel.Result) {
    val codec = call.argument<String>("encoder")

    val isSupported = AudioFormats.isEncoderSupported(
      AudioFormats.getMimeType(codec)
    )

    result.success(isSupported)
  }
}